---
title: "PairSeq Pipeline"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Full PairSeq Workflow

* In Rmd format purely for development purposes
* Once optimized, all code will be packaged into functions

```{r}
library(devtools)
library(PairSeq) # All PairSeq Functions

library(data.table)
library(stringr)
# libraries from PairSeq functions
library(tibble)
library(seqinr)  # write.fasta
library(dplyr)
library(fmsb)  # for percentiles

data_dir     <- "~/rprojects/PairSeq_Development/phage_data/FTD_cohort/"
results_dir  <- "~/rprojects/PairSeq_Development/Results/"

setwd(data_dir)           # Set working directory
dir.create(results_dir,recursive = T)
```

# INPUT FILES
```{r}
meebo_data_fh          <- paste(c(data_dir,"FTD_Full_Phage_Run_w_controls.csv"),collapse = "")
peptide_gene_map_fh    <- paste(c(data_dir,"peptide_gene_mapping.csv"),collapse = "")
background_peptides_fh <- paste(c(data_dir,"background_peptides.txt"),collapse = "")
metadata_fh            <- paste(c(data_dir,"FTD_clinical_sample_names.csv"),collapse = "")
```


# 1. Load peptide map and meebo data

### a. Meebo/peptide table results
* Dementia (),Healthy
```{r}
meebo_data <- read.csv(meebo_data_fh,stringsAsFactors = F)
peptide_gene_map <- read.csv(peptide_gene_map_fh,stringsAsFactors = F)

# List of peptide IDs used for setting RPK threshold
background_peptides <- read.table(background_peptides_fh,stringsAsFactors = F)
background_peptides <- background_peptides$V1
```

### b. sample metadata
```{r}
ftd_sample_names_df <- read.csv(metadata_fh,stringsAsFactors = F)
disease_names <- names(ftd_sample_names_df)
disease_names <- disease_names[2:length(disease_names)]
orig_names    <- as.character(ftd_sample_names_df[1,])
orig_names    <- orig_names[2:length(orig_names)]

sample_conv_list <- list()
for(i in 1:length(orig_names)){
  sample <- orig_names[i]
  sample_conv <- disease_names[i]
  
  sample_conv_list[[sample]] <- sample_conv
}

```


### c. Convert sample names so it has Disease category
```{r}
current_cols <- names(meebo_data)

for (i in 2:length(current_cols)){
  sample_current <- current_cols[i]
  if(grepl("^X",sample_current)){
    sample_current <- substr(sample_current,2,nchar(sample_current))
  }
  
  sample_conv <- ""
  if (!grepl("BEADS_|GPHN_|GFAP_",sample_current)){
    sample_conv    <- sample_conv_list[[sample_current]]
  } else{
    sample_conv <- sample_current
  }
  
  current_cols[i] <- sample_conv
}

names(meebo_data) <- current_cols
```

# 2. Merge both data sets (by peptide column)
```{r}
colnames(peptide_gene_map)[1] <- "peptide"
df <- merge(meebo_data, peptide_gene_map, by = "peptide")
```

# 3. Create unique identifier column
```{r}
df$peptide_id <- paste(df$gene, df$index,sep = "_")
```


# 4. Clinical sample table (establish different comparison groups and Bead controls)
* Establish vector of control and disease columns
* Bead controls = GFAP columns

* Ref control = NC and BEAD (negative controls, background)
* positive controls = GFAP, GPHN
* Disease = all disease samples

```{r}
column_names    <- names(df)
bead_controls   <- column_names[grep("GFAP|GPHN",column_names)]
gfap_controls   <- column_names[grep("GFAP",column_names)]
ref_controls    <- column_names[grep("NC_|BEADS",column_names)]
disease_samples <- column_names[! column_names %in% c("peptide_id",names(peptide_gene_map), bead_controls,ref_controls)]

# CSF - disease/reference
disease_samples_csf     <- disease_samples[grep("CSF",disease_samples)]
disease_samples_plasma  <- disease_samples[! disease_samples %in% disease_samples_csf]

# Plasma Cells - disease/reference
ref_controls_csf     <- ref_controls[grep("CSF|BEADS",ref_controls)]
ref_controls_plasma  <- ref_controls[grep("plasma|BEADS",ref_controls)]
```


# 5. Positive Control Percentile file of peptides
* PUT all of this code into the full parse function. This 
* Calculate percentile of lowest abuntant positive control (GFAP)
* use this percentile to filter full meebo data frame to get rid of low expression peptides
* To start use 15 select GFAP peptides identified

UPDATE:
* There are 4 GFAP samples, take the lowest abundace peptides of the average of these samples and set the percentile of this to be the threshold for the entire dataset
* Alternatively set a hard threshold of 2 RPK
```{r}
# Establish Percentile threshold based on GFAP samples
df_background     <- df[,c("peptide_id","peptide","gene",gfap_controls)]

all_min_gfap_pct <- c()
for(sample in gfap_controls){
  sample_rpk        <- df_background[,sample]
  names(sample_rpk) <- df_background$peptide
  sample_rpk        <- sample_rpk[sample_rpk > 0]
  
  pct_vec        <- percentile(sample_rpk)
  names(pct_vec) <- names(sample_rpk)
  
  pct_vec_target   <- pct_vec[names(pct_vec) %in% background_peptides]
  min_gfap_pct     <- min(pct_vec_target)
  all_min_gfap_pct <- c(all_min_gfap_pct,min_gfap_pct)
}

MIN_PCT_THRESH <- mean(all_min_gfap_pct)


# Apply this threshold to entire data set
all_samples   <- c(bead_controls,ref_controls,disease_samples)

peptide_ids_to_keep <- c()
for(sample in all_samples){
  sample_rpk        <- df[,sample]
  names(sample_rpk) <- df$peptide_id
  # All rpk must be greater than zero
  sample_rpk    <- sample_rpk[sample_rpk > 0]
  
  # Calculate Percentiles
  pct_vec        <- percentile(sample_rpk)
  names(pct_vec) <- names(sample_rpk)

  pct_vec_filt <- pct_vec[pct_vec > MIN_PCT_THRESH]
  peptide_ids_to_keep <- c(peptide_ids_to_keep,names(pct_vec_filt))
}

peptide_ids_to_keep <- unique(peptide_ids_to_keep)

# Filter out all petides with low RPK
df_filt <- df[df$peptide_id %in% peptide_ids_to_keep,]
```

# 6. calcluating ref mean on Filtered data
```{r}
# Ctrl Means
df_filt$mean_ref_csf    <- apply(df_filt[,c(ref_controls_csf)], 1, mean)     # Control group
df_filt$mean_ref_plasma <- apply(df_filt[,c(ref_controls_plasma)], 1, mean)  # Control group

df_filt$mean_pos     <- apply(df_filt[,c(bead_controls)], 1, mean)    # Bead Control group
df_filt$mean_exp     <- apply(df_filt[,c(disease_samples)], 1, mean)  # Disease group
```



# 7. Caluculate FC (PairSeq_make_FC_df.r) - adds FC columns
* FC is calculated for disease group and positive controls (bead controls)
* Normalize (divide by mean) disease and bead samples by reference controls
* Calculate FC for CSF and Plasma separately then merge
```{r}
# PairSeq_make_FC_df.r  (one function)
df_fc_csf_exp <- makeFCdf(df_filt,
                 target_columns = disease_samples_csf,
                 mean_column    = "mean_ref_csf",  onlyFC = T)
df_fc_plasma_exp <- makeFCdf(df_filt,
                 target_columns = disease_samples_plasma,
                 mean_column    = "mean_ref_plasma", onlyFC = T)

df_fc_exp <- cbind(df_filt,df_fc_csf_exp,df_fc_plasma_exp)

#df_fc_ref <- makeFCdf(df_filt,
#                 target_columns = ref_controls,
#                 mean_column    = "mean_exp")
#df_fc_pos <- makeFCdf(df_filt,
#                 target_columns = bead_controls,
#                 mean_column    = "mean_pos")

```


# 8. (PairSeq_full_parse_fasta.r) - Apply all FC/RPK filters
* This step is meant to apply all FC and RPK filters to narrow down a set of candidates
```{r}
start.time <- Sys.time()

# Use a subset of data frame to optimize speed
candidate_df <- fullParse(df = df_fc_exp,list_of_samples = disease_samples,FC_THRESH2 = 100)

save(candidate_df,file = paste(c(data_dir,"final_candidates.RData"),collapse = ""))

end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken

#full_parse(df = df_fc_ref,list_of_samples = ref_controls,FC_THRESH2 = 34) # 34 FC is the Max

# full_parse(df = df_fc_pos,list_of_samples = bead_controls)
```



# 9. Use Kmer analysis to further filter candidates
* keeping peptides that have at least a 7 AA sequence overlap.
* Extend KMER analysis past the gene level to consider all genes
```{r}
start.time <- Sys.time()

candidate_df_fin <- findKmerOverlap.old(candidate_df,KMER_SIZE = 7)
save(candidate_df,candidate_df_fin,file = "final_candidates.RData")


end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken

write.csv(candidate_df_fin,file = "~/rprojects/PairSeq_Development/Results/FTD_candidate_peptides.csv",quote = F)
```


### --take small subset of phage data to optimize speed of fullParse function
```{r}
# only keep a subset of genes
target_genes <- unique(candidate_df$gene)
target_genes <- target_genes[1:100]

candidate_df_sub <- candidate_df[candidate_df$gene %in% target_genes,]
```

### -- RUN kmer analysis on subset
*  OPTIMIZE for speed and add in new columns to help with ranking
```{r}
start.time <- Sys.time()

candidate_df_sub_fin <- findKmerOverlap.old(candidate_df_sub,KMER_SIZE = 7)

end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
```

# 9.5 ALTERNATIVE approach
* try doing a mutliple alignment follow by kcount

### a. Create output FASTA
```{r}
library(seqinr)

# Subset results
target_genes     <- unique(candidate_df$gene)
target_genes     <- target_genes[1:200]
candidate_df_sub <- candidate_df[candidate_df$gene %in% target_genes,]

# write FASTA
fh <- "~/rprojects/PairSeq_Development/candidate_peptides.subset.fasta"
write.fasta(sequences = as.list(candidate_df$sequence),names = as.list(candidate_df$peptide_id),file.out = fh)

```

### b. kcount
```{r}
library(kmer)
library(ape)
fh               <- "~/rprojects/PairSeq_Development/candidate_peptides.subset.fasta"
mySequences      <- read.FASTA(fh,type = "AA")

# Kcount of 7mers
start.time <- Sys.time()

kmer_count_table <- kcount(mySequences,k = 7)

end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
```

#### --SAVE ENV
```{r}
save(list = ls(),file = "~/rprojects/PairSeq_Development/currentEnv.RData")
```

### c. Identify all peptides that share Kmers
```{r}
start.time <- Sys.time()

pep_to_keep      <- list()
pep_gene_overlap <- list()
for (pep in row.names(kmer_count_table)) {
  kmer_df <- kmer_count_table[row.names(kmer_count_table) %in% pep,]
  target_kmers <- names(kmer_df[kmer_df > 0])
  
  kmer_df_sub <- kmer_count_table[,colnames(kmer_count_table) %in% target_kmers]
  kmer_df_sub <- as.data.frame(kmer_df_sub) 
  # Change the summing somehow
  kmer_df_sub$total <- rowSums(kmer_df_sub)
  kmer_df_sub       <- kmer_df_sub[kmer_df_sub$total > 0,]
  
  if (nrow(kmer_df_sub) > 1){
    total_overlapping_pep <- length(unique(row.names(kmer_df_sub))) -1 
    pep_to_keep[[pep]] <- total_overlapping_pep
    
    # Check for gene level overlap
    pep_gene <- unlist(strsplit(pep,"_"))[1]
    kmer_df_sub$gene <- tstrsplit(row.names(kmer_df_sub),"_")[[1]]
    kmer_df_gene <- kmer_df_sub[kmer_df_sub$gene %in% pep_gene,]
    if (nrow(kmer_df_gene) > 1){
      total_overlapping_pep_gene <- length(unique(row.names(kmer_df_gene))) -1 
      pep_gene_overlap[[pep]] <- total_overlapping_pep_gene
    }
    
  } else{
    # No Kmer overalp
    cat(paste0(pep,"\n"))
  }
  
}

end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
```

### d. Subset final candidate data frame
```{r}
candidate_df_fin <- candidate_df[candidate_df$peptide_id %in% names(pep_gene_overlap),]
```




# 10. Mathamatically assign a significant ranking system for final peptide results (Can wait until after PairSeq development is complete)
* RPK, FC, complexity of KMers
* create a Kmer count table of human proteome 
* If petides span muliple isoforms of the same gene, apply a ratio weight so isoform with the higher RPK getting a higher weight

Validation sets on MEEBO:
* Anti hu & anti Yo
* Sarah Vasquez Auto antibody sets

### a. Try to assess the distribution and stats
```{r}
library(ggplot2)
disease_samples_fc <- disease_samples

df_subset <- candidate_df_fin[,names(candidate_df_fin) %in% disease_samples]

df_subset$mean <- rowMeans(df_subset)
df_subset <- df_subset[df_subset$mean > 0,]

p <- ggplot(df_subset, aes(x=mean)) + geom_density()
p
```


### b. ranking via random forrest
* Trying to follow some online material to learn how impliment balanced random forrest
https://stats.stackexchange.com/questions/171380/implementing-balanced-random-forest-brf-in-r-using-randomforests/373695
  * random forrest tutorial
  https://towardsdatascience.com/random-forest-in-r-f66adf80ec9

* Alternatively we could try SVM
```{r}
library(ranger) #Best random forest implementation in R

#Make a mock dataset 
set.seed(43)
nrow <- 1000
ncol <- 10
X <- matrix(rnorm(nrow * ncol), ncol=ncol)
CF <- rnorm(ncol)
Y <- (X %*% CF + rnorm(nrow))[,1]
Y <- as.integer(Y > quantile(Y, 0.90))
table(Y)

#Compute weights to balance the RF
w <- 1/table(Y)
w <- w/sum(w)
weights <- rep(0, nrow)
weights[Y == 0] <- w['0']
weights[Y == 1] <- w['1']
table(weights, Y)

#Fit the RF
data <- data.frame(Y=factor(ifelse(Y==0, 'no', 'yes')), X)
model <- ranger(Y~., data, case.weights=weights)
print(model)

```



# 11. Post analysis filter
* feed peptide list into a database that reveals pepides with experimenatlly validated significance (example: peptides that show significant staining in lung tissue of patients with MS)

* This post analysis could be an interactive shiny app

* Rank score: rpk, FC, number of peptides it overlaps

* query databases to pull info on each gene 

```{r}
library(hpar)
data(hpaNormalTissue)

# Overlapping genes
overlapping_genes <- unique(candidate_df$gene[candidate_df$gene %in% hpaNormalTissue$Gene.name])

# Test
hpaNormalTissue[hpaNormalTissue$Gene.name %in% overlapping]

```



